### ISSUE TITLE
Service account per Job

### ISSUE THREAD

**Issue by mumoshu** [Azure/brigade#251]
I'd wanted to associate a service account with enough permissions to deploy things to k8s, only to the job runs a k8s deployment.

However, as far as I can see from code, [brigade seems to hard-code `brigade-worker` serviceaccount to be used by all the job runner pod](https://github.com/Azure/brigade/blob/8346887653a41ca7879801e64d1f3f20561273b6/brigade-worker/src/k8s.ts#L585). It would be great if it could be overridden like `job.serviceaccount = "deployer"` in brigade.js for this use-case.

**Comment by wstrange** [comment]
I think that running kubectl , Helm, etc. from within Jobs will be very common.

**Comment by mumoshu** [comment]
Thanks for the response!

Regarding kubectl, Yes - that's why this could be important.

Not sure about helm though. AFAIK tiller rather than helm is responsible to manage k8s resources. The serviceaccout associated to the pod running helm won't be used by tiller(at least for now). We'd need to have tiller-per-ns instead.

**Comment by technosophos** [comment]
The one thing I am worried about is whether that would open up a security hole. Are there cases where the script could then grant a Job a higher level of permissions then it is supposed to have?

I am not entirely clear about what restrictions Kubernetes puts on the serviceaccount part of a pod definition. If we can do this without the risk of a script being able to escalate privileges, I'd be happy to implement this feature.

**Comment by mumoshu** [comment]
@technosophos Thanks for the response!

> I am not entirely clear about what restrictions Kubernetes puts on the serviceaccount part of a pod definition. 

AFAIK, RBAC can't be used for restricting which serviceaccount can be referred via the `serviceAccountName` field in a pod spec. An user who has permissions to "create pods" is able to mount any serviceaccount token to a pod. Try:

```yaml
# rbac.yaml
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: RoleBinding
metadata:
  name: foo-pod-all
subjects:
- kind: ServiceAccount
  name: foo
roleRef:
  kind: ClusterRole
  name: pod-all
  apiGroup: rbac.authorization.k8s.io
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRole
metadata:
  name: pod-all
rules:
- apiGroups:
  - ""
  resources:
  - pods
  - pods/attach
  verbs:
  - create
  - delete
  - deletecollection
  - get
  - list
  - patch
  - update
  - watch
- apiGroups:
  - ""
  resources:
  - pods/log
  - pods/status
  verbs:
  - get
  - list
  - watch
```

```
$ ns=yourns

$ alias k="kubectl --namespace $ns"

$ another_sa=<any sa shown by running `kubectl get sa`>

$ k create sa foo

$ k apply -f rbac.yaml

$ k  run --restart=Never --as system:serviceaccount:$ns:foo --image alpine:edge -it --overrides="$(jq -n '{spec:{serviceAccountName:"$another_sa"}}')" mytestpod -- uptime
rolebinding "foo-pod-all" configured
clusterrole "pod-all" configured
 04:10:35 up 76 days, 21:14,  load average: 0.41, 0.16, 0.11

# verify that the serviceaccountname does exist in the pod spec
$ k get po mytestpod -o yaml
```

>  If we can do this without the risk of a script being able to escalate privileges

Assuming RBAC doesn't work in this case as above, I guess a whitelist of allowed serviceaccount per brigade project, which is an empty list by default, would work.

```yaml
allowedServiceAccounts:
- deployer-sa
```

```js
events.on('push', (e, p) => {
  const job = new Job(...)
  job.serviceAccount = "deployer-sa"
  job.run()
})
```

**Comment by mumoshu** [comment]
BTW, when in a case that anyone other than you(as an admin) can freely create brigade projects, my suggested solution won't prevent a specific brigade project from whitelisting undesirable serviceaccount(s).

If anyone that case, I'd suggest extending my above solution further by allow whitelisting serviceaccounts at brigade-installation level.

Say we have this in a brigade's(not a brigade project) config:

```yaml
allowedServiceAccounts:
- reader-sa
```

No brig user would be able to white-list `deployer-sa` in a brigade project.

**Comment by mumoshu** [comment]
I also believe brigade shouldn't allow the user to change the serviceaccount used by a job pod by default, so that there would be no security hole by default.
After that, it is all up to the user - we should at least warn not to give too much permissions to the custom serviceaccount?

**Comment by mumoshu** [comment]
Update: Just realized that it seems like I can modify anything by directly using `JobRunner` or `kubernetes.v1Pod`+`kubernetes.Core_v1Api` client instead of Job.

Then, introducing lists of allowed service account names like I've suggested doesn't add real benefits to security, as you can basically do anything w/ what is accessible to the runner's nodejs runtime.

## TL;DR; What would be an alternative approach?

I believe we can safely allow setting `job.serviceAccountName` without any validation, as long as the user deploys 1 brigade per ns per team/project/github repo/etc while making the ns dedicated to the brigade cluster.

## Background

### Every brigade component (as of today) seems to expect to be deployed in the same ns

The reasoning behind this is that currently it seems like all of:

- brigade gateway
- brigade controller
- brigade worker
- brigade job runner

MUST reside in a same namespace. Otherwise, various bad things could happen:

- A controller can't access a build(persisted as a k8s secret) from gateway
- A worker can't mount secrets for the build and the project
- A job runner can't mount the secrets containing imagePullSecret and the SSH key

### Protecting your k8s secrets from (potentially malicious) brigade scripts

RBAC should be used to restrict brigade worker to create job runner pods in a "safe" ns.
A "safe" ns is one which contains secrets only relevant to specific project/team.
Note that "secrets" includes:

- custom secrets could be mounted to job runner pods
- secrets representing brigade projects, builds, etc.

### The default RBAC role/binding can be used as-is

Brigade's default worker role defined in `charts/brigade/templates/worker-role.yaml` already restrict any worker to the ns a brigade project resides. So you don't need to tweak roles & bindings here.

## Again: the alternative approach

You just need to either of:

1. Ensure the `default` ns to be dedicated to brigade or
2. More practically, explicitly set ns to be something other than `default`, dedicated to specific brigade/brigade project/team/github repo/etc.

And:

- Any custom service account should have properly restricted roles & bindings. Note that, for example, a custom service account has a ClusterRoleBinding to permit creating pod in ANY namespace results in a potential leakage of ANY secret in ANY namespace.

to protect your secrets.

We can then add `job.serviceAccountName` for an "easier" way to associate one of service accounts in the namespace.

**Comment by technosophos** [comment]
Thanks. Yes, I think that you are correct: If Brigade and RBAC are configured properly, then that's about as secure as we can get things. So whitelisting is not necessary.

Adding this to the list of things to do.
